src/klein/_request.py

```diff
@@ -0,0 +1,341 @@
+        # type: () -> bytes
+        if type(self._body) is bytes:
+            return succeed(self._body)
+
+        # assuming: IFount.providedBy(self._body)
+
+        def cache(bodyBytes):
+            # type: (bytes) -> bytes
+            self._state._cachedBody = bodyBytes
+            return bodyBytes
+
+        d = fountToBytes(self._body)
+        d.addCallback(cache)
+        return d
+
+
+
+# Support for L{IRequest}
+
+@implementer(IHTTPRequest)
+@attrs(frozen=True)
+class HTTPRequestFromIRequest(object):
+    """
+    HTTP request.
+
+    This is an L{IHTTPRequest} implementation that wraps an L{IRequest} object.
+
+    This is used by Klein to expose objects from L{twisted.web} to clients
+    while presenting the Klein interface.
+    """
+
+    @attrs(frozen=False)
+    class _State(object):
+        """
+        Internal mutable state for L{HTTPRequestFromIRequest}.
+        """
+
+        _cachedBody = attrib(
+            validator=optional(instance_of(bytes)), default=None, init=False
+        )  # type: bytes
+
+    _request = attrib(validator=provides(IRequest))       # type: IRequest
+
+    _state = attrib(default=Factory(_State), init=False)  # type: _State
+
+
+    @property
+    def method(self):
+        # type: () -> Text
+        return self._request.method.decode("ascii")
+
+
+    @property
+    def uri(self):
+        # type: () -> URL
+        request = self._request
+
+        # This code borrows from t.w.server.Request._prePathURL.
+
+        if request.isSecure():
+            scheme = u"https"
+        else:
+            scheme = u"http"
+
+        netloc = nativeString(request.getRequestHostname())
+
+        port = request.getHost().port
+        if request.isSecure():
+            default = 443
+        else:
+            default = 80
+        if port != default:
+            netloc += u":{}".format(port)
+
+        path = nativeString(request.uri)
+        if path and path[0] == u"/":
+            path = path[1:]
+
+        return URL.fromText(u"{}://{}/{}".format(scheme, netloc, path))
+
+
+    @property
+    def headers(self):
+        # type: () -> IHTTPHeaders
+        return HTTPHeadersFromHeaders(headers=self._request.requestHeaders)
+
+
+    def bodyAsFount(self):
+        # type: () -> IFount
+        source = self._request.content
+        if source is None:
+            raise NoContentError()
+
+        fount = IOFount(source=source)
+
+        self._request.content = None
+
+        return fount
+
+
+    def bodyAsBytes(self):
+        # type: () -> bytes
+        if self._state._cachedBody is not None:
+            return succeed(self._state._cachedBody)
+
+        def cache(bodyBytes):
+            self._state._cachedBody = bodyBytes
+            return bodyBytes
+
+        fount = self.bodyAsFount()
+        d = fountToBytes(fount)
+        d.addCallback(cache)
+        return d
+
+
+
+# Fount-related utilities that probably should live in tubes.
+
+def fountToBytes(fount):
```

Comment:
I'm trying to avoid adding "tubes maintainer" to my portfolio, but here's a pair of tickets for these:
 * https://github.com/twisted/tubes/issues/60
 * https://github.com/twisted/tubes/issues/61